java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -254,51 +224,95 @@ class DatabaseImpl(
         }
     }
 
-    private fun getEntityFieldValue(
-        fieldValueId: FieldValueId,
-        field: SchemaField,
-        db: SQLiteDatabase,
-        counters: Counters?
-    ): Any = when {
-        field.isCollection -> {
-            counters?.increment(DatabaseCounters.GET_ENTITY_FIELD_VALUE_COLLECTION)
-            getCollectionEntries(fieldValueId, field.typeId, db, counters)
-        }
-        isPrimitiveType(field.typeId) -> {
-            counters?.increment(DatabaseCounters.GET_ENTITY_FIELD_VALUE_PRIMITIVE)
-            getPrimitiveValue(fieldValueId, field.typeId, db, counters)
-        }
-        else -> {
-            counters?.increment(DatabaseCounters.GET_ENTITY_FIELD_VALUE_REFERENCE)
-            getReferenceValue(fieldValueId, db)
+    /** Returns the singleton and collection fields for the entity at the given storage key ID. */
+    private fun getEntityFields(
+        storageKeyId: StorageKeyId,
+        counters: Counters?,
+        db: SQLiteDatabase
+    ): Pair<Map<FieldName, Referencable?>, Map<FieldName, Set<Referencable>>> {
+        counters?.increment(DatabaseCounters.GET_ENTITY_FIELDS)
+        val singletons = mutableMapOf<FieldName, Referencable?>()
+        val collections = mutableMapOf<FieldName, MutableSet<Referencable>>()
+
+        db.rawQuery(
+            """
+                SELECT
+                    t.name,
+                    t.is_collection,
+                    t.type_id,
+                    t.value_id,
+                    text_primitive_values.value,
+                    number_primitive_values.value,
+                    entity_refs.entity_id,
+                    entity_refs.backing_storage_key,
+                    entity_refs.version_map,
+                    entity_refs.creation_timestamp,
+                    entity_refs.expiration_timestamp
+                FROM (
+                    SELECT
+                        fields.name,
+                        fields.type_id,
+                        fields.is_collection,
+                        CASE
+                            WHEN fields.is_collection = 0 THEN field_values.value_id
+                            ELSE collection_entries.value_id
+                        END AS value_id
+                    FROM storage_keys
+                    LEFT JOIN entities
+                        ON entities.storage_key_id = storage_keys.id
+                    LEFT JOIN fields
+                        ON fields.parent_type_id = storage_keys.value_id
+                    LEFT JOIN field_values
+                        ON field_values.entity_storage_key_id = storage_keys.id
+                        AND field_values.field_id = fields.id
+                    LEFT JOIN collection_entries
+                        ON fields.is_collection = 1
+                        AND collection_entries.collection_id = field_values.value_id
+                    WHERE storage_keys.id = ?
+                ) AS t
```

Comment:
Can this be done without an inner select?

Reply:
Good call! I didn't realise you could reference fields constructed in the SELECT query in a JOIN, but you totally can! This is much nicer now.